### PR TITLE
[REEF-1403] Deadlock between ContextRuntime.StartTask and HeartBeatMa…

### PR DESCRIPTION
**Pull request by afchung**
…nager.OnNext(Alarm)

This addressed the issue by
  * Locking on HeartBeatManager on Evaluator startup in EvaluatorRuntime.

JIRA:
  [REEF-1403](https://issues.apache.org/jira/browse/REEF-1403)

### CODE REVIEW

**Comment by markusweimer** [comment]
I'll test this.

**Comment by markusweimer** [comment]
The tests pass for me, but fail on AppVeyor. @afchung can you check the appveyor logs?

**Comment by afchung** [comment]
@markusweimer detected another possible deadlock and fixed it, please have another look. Thanks!

**Comment by afchung** [comment]
@markusweimer AppVeyor is failing on the known failing test.

**Comment by markusweimer** [comment]
I'll test this.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorRuntime.cs [hunk 1]
```diff
@@ -155,18 +155,21 @@ public void Handle(EvaluatorControlProto message)
...
-            Logger.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "Evaluator state: {0}", _state));
-            EvaluatorStatusProto evaluatorStatusProto = new EvaluatorStatusProto
+            lock (_heartBeatManager)
             {
-                evaluator_id = _evaluatorId,
-                state = _state
-            };
-            return evaluatorStatusProto;
+                Logger.Log(Level.Verbose, "Evaluator state: {0}", _state);
+                EvaluatorStatusProto evaluatorStatusProto = new EvaluatorStatusProto
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorRuntime.cs]: Maybe `return new ...`?

**Comment by markusweimer** [comment]
> AppVeyor is failing on the known failing test.

Dang! I was hoping the fix for this would also fix that one.

**Comment by afchung** [comment]
@markusweimer are we able to get the logs for that test failure? I was never able to reproduce it locally.

**Comment by markusweimer** [comment]
I now have a failure in `Org.Apache.REEF.Tests.Functional.Failure.TestEvaluatorWithActiveContextImmediatePoison.TestPoisonedActiveContextHandlerImmediate`. I will locate the logs and make them available.

**Comment by afchung** [comment]
@markusweimer Addressed your comment. Thanks!

**Comment by markusweimer** [comment]
Logs are [here](https://onedrive.live.com/redir?resid=5801726772BFC3DA!287230&authkey=!AGRkh8VgXznoaHQ&ithint=file%2czip)

**Comment by afchung** [comment]
@markusweimer I've taken a brief look at the test, and I believe that the test is failing on a race condition. The test triggers an `Alarm` on `ContextStart` which fails the Evaluator and exits the Evaluator process. On the exit of the Evaluator process, the local ResourceManager detects that the process has exited first and tells the Driver that the Evaluator has failed before the first Context Heartbeat is triggered, so the Driver does not know about the Evaluator's failed context, which our test asserts to exist, causing the test failure. We should fix the test such that we check for a failed context only after the `ActiveContextHandler` has been triggered which will need to rely on Driver-Evaluator messaging.

**Comment by afchung** [comment]
I'm not particularly fond of that fix though, since it adds additional complexity into what is supposed to be a fairly simple test.

**Comment by markusweimer** [comment]
OK. So this change can go in. Will merge.

**Comment by markusweimer** [comment]
> The test triggers an Alarm on ContextStart which fails the Evaluator and exits the Evaluator process.

I think that is the bug. The Evaluator should first send the error to the Driver and then exit.

**Pull request closed without merging**